tools/test_modules/m27000.pm

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env perl
+
+##
+## Author......: See docs/credits.txt
+## License.....: MIT
+##
+
+use strict;
+use warnings;
+
+use Digest::HMAC qw (hmac hmac_hex);
+use Digest::MD5  qw (md5);
+use Encode       qw (encode);
+use Crypt::ECB   qw (encrypt);
+
+sub setup_des_key
+{
+  my @key_56 = split (//, shift);
+
+  my $key = "";
+
+  $key = $key_56[0];
+
+  $key .= chr (((ord ($key_56[0]) << 7) | (ord ($key_56[1]) >> 1)) & 255);
+  $key .= chr (((ord ($key_56[1]) << 6) | (ord ($key_56[2]) >> 2)) & 255);
+  $key .= chr (((ord ($key_56[2]) << 5) | (ord ($key_56[3]) >> 3)) & 255);
+  $key .= chr (((ord ($key_56[3]) << 4) | (ord ($key_56[4]) >> 4)) & 255);
+  $key .= chr (((ord ($key_56[4]) << 3) | (ord ($key_56[5]) >> 5)) & 255);
+  $key .= chr (((ord ($key_56[5]) << 2) | (ord ($key_56[6]) >> 6)) & 255);
+  $key .= chr (( ord ($key_56[6]) << 1) & 255);
+
+  return $key;
+}
+
+sub get_random_netntlmv1_salt
+{
+  my $len_user   = shift;
+  my $len_domain = shift;
+
+  my $char;
+  my $type;
+  my $user = "";
+
+  for (my $i = 0; $i < $len_user; $i++)
+  {
+    $type = random_number (1, 3);
+
+    if ($type == 1)
+    {
+      $char = random_numeric_string (1);
+    }
+    elsif ($type == 2)
+    {
+      $char = random_uppercase_string (1);
+    }
+    else
+    {
+      $char = random_lowercase_string (1);
+    }
+
+    $user .= $char;
+  }
+
+  my $domain = "";
+
+  for (my $i = 0; $i < $len_domain; $i++)
+  {
+    $type = random_number (1, 3);
+
+    if ($type == 1)
+    {
+      $char = random_numeric_string (1);
+    }
+    elsif ($type == 2)
+    {
+      $char = random_uppercase_string (1);
+    }
+    else
+    {
+      $char = random_lowercase_string (1);
+    }
+
+    $domain .= $char;
+  }
+
+  my $c_challenge = random_bytes (8);
+  my $s_challenge = random_bytes (8);
+
+  my $salt_buf = $user . "::" . $domain . ":" . unpack ("H*", $c_challenge) . unpack ("H*", $s_challenge);
+
+  return $salt_buf;
+}
+
+sub module_constraints { [[32, 32], [-1, -1], [32, 32], [-1, -1], [-1, -1]] } # room for improvement in pure kernel mode
```

Comment:
There's no optimized kernel. This confused test.pl in single and passthrough mode. I think you want something like this:

```
sub module_constraints { [[32, 32], [-1, -1], [-1, -1], [-1, -1], [-1, -1]] } # room for improvement in pure kernel mode
```

The first two define the pure kernel, the other three the optimized kernel. The same issue is in m27100.pm